tests/test_request_id.py

```diff
@@ -1,77 +1,398 @@
+from flask import request
+from itertools import chain, product
 import mock
-from werkzeug.test import EnvironBuilder
+import pytest
 
-from dmutils import request_id
-from dmutils.request_id import CustomRequest
+from dmutils.request_id import init_app as request_id_init_app
 
 
-def test_get_request_id_from_request_id_header():
-    builder = EnvironBuilder()
-    builder.headers['DM-REQUEST-ID'] = 'from-header'
-    builder.headers['DOWNSTREAM-REQUEST-ID'] = 'from-downstream'
-    request = CustomRequest(builder.get_environ())
+_GENERATED_TRACE_VALUE = "di5ea5e5deadbeefbaadf00dabadcafe"
 
-    request_id = request._get_request_id('DM-REQUEST-ID',
-                                         'DOWNSTREAM-REQUEST-ID')
 
-    assert request_id == 'from-header'
+_trace_id_related_params = (
+    (
+        # extra_config
+        {
+            "DM_REQUEST_ID_HEADER": "DM-REQUEST-ID",
+            "DM_DOWNSTREAM_REQUEST_ID_HEADER": "DOWNSTREAM-REQUEST-ID",
+        },
+        # extra_req_headers
+        (
+            ("DM-REQUEST-ID", "from-header",),
+            ("DOWNSTREAM-REQUEST-ID", "from-downstream",),
+        ),
+        # expected_trace_id
+        "from-header",
+        # expect_uuid_call
+        False,
+        # expected_onwards_req_headers
+        {
+            "DM-REQUEST-ID": "from-header",
+            "DOWNSTREAM-REQUEST-ID": "from-header",
+        },
+        # expected_resp_headers
+        {
+            "DM-REQUEST-ID": "from-header",
+            "DOWNSTREAM-REQUEST-ID": "from-header",
+        },
+    ),
+    (
+        {
+            "DM_REQUEST_ID_HEADER": "DM-REQUEST-ID",
+            "DM_DOWNSTREAM_REQUEST_ID_HEADER": "DOWNSTREAM-REQUEST-ID",
+        },
+        (
+            ("DOWNSTREAM-REQUEST-ID", "from-downstream",),
+        ),
+        "from-downstream",
+        False,
+        {
+            "DM-REQUEST-ID": "from-downstream",
+            "DOWNSTREAM-REQUEST-ID": "from-downstream",
+        },
+        {
+            "DM-REQUEST-ID": "from-downstream",
+            "DOWNSTREAM-REQUEST-ID": "from-downstream",
+        },
+    ),
+    (
+        {
+            "DM_REQUEST_ID_HEADER": "DM-REQUEST-ID",
+            "DM_DOWNSTREAM_REQUEST_ID_HEADER": "",
+        },
+        (),
+        _GENERATED_TRACE_VALUE,
+        True,
```

Comment:
I think one way of making these be more readable would be if there were more dictionaries instead of tuples - without a lot of scrolling it's not clear what this `True` thing actually is/means (for example).

Reply:
Problem with dictionaries is it means you've got to write a lot more boilerplate *everywhere* and it introduces more uncertainty - possibilities of missing keys, and simply the fact that dictionaries are mutable gives the reader more things they have to take into account when reading it. If you really just want more labelling where these tuples are defined, I *can* just repeat the labelling comments each time.

Reply:
Sure, yep - they would take less space if they were on the same line as the thing they were commenting

Reply:
Don't want to do with battle with flake8 over *that* one...